Full Trade Quest Shuffle and Granny's Potion Shop Item Shuffle

mracsys opened the pull request:
Allow shuffling all child and adult trade items in the pool simultaneously. Additionally, Shuffle Medigoron and Carpet Salesman is refactored to Shuffle Expensive Merchants to include the Blue Potion sale in Granny's Potion Shop after trading in the Odd Mushroom.

Fix #14, #1239, #1577

Use two sets of unused scene flags to indicate trade item owned and traded status. These are independent of existing game flags set during the trade sequence. These new flags allow the following:

1. Adult trade items cannot be turned in multiple times.
   EXCEPTION: Claim Check can be presented to Biggoron multiple times in vanilla. This behavior is preserved.
2. Trade items can be cycled in the inventory using the d-pad.
![image](https://user-images.githubusercontent.com/19615534/190527554-7787fc52-6d0c-4a9c-a0f3-4a70fc3812ad.png)

The following additional changes are made:

* Trade Item owned flags are set in the get item upgrade callback
* Trade Item traded flags are set in the get item override function before the item is overridden, assuming the incoming item ID - 1 for the traded item ID.
* Dungeon info screens cannot be displayed if the cursor is over a shuffled trade slot to prevent conflicts. If the trade slot is not fully shuffled, the screens are accessible.
* Full adult trade shuffle disables trade item reverts and timers
* All adult trade items deselected in the Adult Trade Sequence Items multiselect will place the Pocket Egg in its vanilla location on Gift from Anju.
    * Previous behavior from #1692 was to randomly pick from all trade items for the shuffled adult trade item. Since this use case is already handled by selecting all items in the multiselect, enable the opposite with no trade items shuffled.
* Adult trade items can be starting items in both the GUI and plando.
    * Error messages are provided for too many adult trade items in starting items and for starting trade items that aren't shuffled.
* Odd Potion can be selected as a shuffled trade item now that Fado and Grog spawns are handled with the new trade flags.
* Fado (Kokiri girl in Lost Woods accepting the Odd Potion) is moved to the side and rotated slightly to allow her to coexist with Grog (Cojiro's owner)
* Both Fado and Grog only spawn in Lost Woods when their respective trade items are owned, but not traded.
    * Fado used to spawn only when the adult trade slot had the Odd Potion, regardless of trade flags
    * Grog used to always spawn until the Odd Mushroom was turned in to Granny.
* The Skull Kid near Fado and Grog can now spawn if neither Fado nor Grog spawn
    * Vanilla behavior was to spawn only when the adult trade slot was empty or greater than or equal to the Poacher's Saw
* The Fado, Grog, and Skull Kid changes apply even if full trade quest shuffle is off. In entrance rando, the trade sequence up to the Saw may not be accessible or only accessible late game. This change allows the Skull Kid to be a more viable rupee farming spot even if a trade item before Saw is found.
* Biggoron will always use his irritated eyes animation until given the Eyedrops
    * Vanilla behavior transitioned to this animation only if the adult trade slot was between Broken Sword and Claim Check.
    * Active even if full trade quest shuffle is off for simplicity
* Granny will sell a blue potion or shuffled item anytime after the Odd Mushroom is traded, even if the active trade item is the Odd Mushroom or earlier.
* Previous fixes for Biggoron and Granny's Potion Shop are replaced
* Child trade shuffle replaces Weird Egg shuffle and Skip Child Zelda
    * Weird Egg shuffle is accomplished by checking the box to shuffle the Weird Egg.
    * Skip Child Zelda requires Zelda's Letter as a starting item and unshuffled.
* Child trade shuffle does not have the same single item shuffle that adult can enable. Adult trade items are generally a one time trade with one NPC. Most child trade items are repeat visits to the mask shop, where previous items are available to take. Adult can assume that a later trade item on Anju locks out earlier trade items. Child cannot make that assumptiom. For example, if Skull Mask were chosen as the shuffled child trade item on Malon, either the mask would still be available in the mask shop or an additional shuffled item on the shop slot would need to be added to prevent early access to Skull Mask. Rather than try to handle this, child trade items are one-for-one when shuffled, similar to full adult trade shuffle.
* Right side masks are not visible until trade quest completion in vanilla except for Mask of Truth. A miscellaneous hints option is added to always show these items. Since they can be peeked visually, none of the right side masks get always-type gossip stone hints.
* Hyrule Castle guards are changed to never block access to Zelda after visiting her, making Zelda's Letter and Song from Impa repeatable. Both locations prevent duping the item if visited multiple times.
* Zelda's Letter will never revert to the Chicken.
* Glitched logic always has the child trade d-pad menu enabled to work around softlocks with Zelda's Letter. This also effectively ensures child will have a repeatable cutscene item after finding any of Weird Egg, Chicken, or Zelda's Letter.

I am really unsure of the changes I made in Plandomizer.py. I tried to preserve old behavior where appropriate. Refactoring child trade item pool behavior is the shakiest part IMO.

An unrelated fix for #1577 is added to allow shuffling the blue potion sale with other shop items via plando. This was originally planned for shuffling the item Granny sells, but I decided to limit Granny shuffle to major items with the other expensive merchants. The bugfix is kept at the request of plando authors. Unused text IDs are overwritten with description and purchase text for blue potions.

r0bd0g commented:
"The Skull Kid near Fado and Grog can now spawn if neither Fado nor Grog spawn"
To be clear the skull kid spawning stuff applied to all skull kids in LW.

"Granny will sell a blue potion anytime after the Odd Mushroom is traded, even if the active trade item is the Odd Mushroom or earlier."
I'm not entirely clear on how this whole shuffle works. Just, make sure that regardless of settings there's always some way to enable buying blue potion (outside of RO making it impossible to perform).

I suggest putting in the work to retain the vanilla Biggoron behaviour when trade shuffle is off. There's no benefit to removing it like you could argue there is with the Skull Kid change, and I like the flavour of him going through his usual phases okay.

mracsys commented:
> "The Skull Kid near Fado and Grog can now spawn if neither Fado nor Grog spawn" To be clear the skull kid spawning stuff applied to all skull kids in LW. If one of them couldn't spawn, none of them could.

The change applies to all of them.

> "Granny will sell a blue potion anytime after the Odd Mushroom is traded, even if the active trade item is the Odd Mushroom or earlier." I'm not entirely clear on how this whole shuffle works. Just, make sure that regardless of settings there's always some way to enable buying blue potion (outside of RO making it impossible to perform).

Blue potions can only be bought if the Odd Mushroom is traded to Granny, whether or not Odd Potion is shuffled. This reminds me I need to handle "normal" race settings where the earliest trade item is prescription. I had that handled when it was two dropdowns for earliest and latest item, but had to take that logic out for the new multiselect. It's fixed again.

> I suggest putting in the work to retain the vanilla Biggoron behaviour when trade shuffle is off. There's no benefit to removing it like you could argue there is with the Skull Kid change, and I like the flavour of him going through his usual phases.

Text behavior should be the same as it was with the old hacks. The animation hacks are in several places that made it messy for something most people will never see. I updated the behavior to be untouched for trade shuffle off, and only change animation with trade shuffle on if any trade items are owned which would have triggered the change in vanilla (between Broken Sword and Eyedrops inclusive).

> I'm not sure if you've done this or not but I suggest an intermediate setting that keeps the timers (when applicable), by retaining the usual cojiro->mushroom and prescription->frog->eyedrops chains. This will give you as many extra checks as possible without removing the timed fetch quest gameplay. (I'm not sure how selecting your possible starting items would interact with this though. I'm not convinced there even should be a starting item setting when trade quest rework is on? I'm curious in general how your PR and the starting item settings interact.)

I did not add this because I wanted to be consistent with the other settings that can affect timers. If a trade item is shuffled it can't revert because it's vanilla location has something else. If it's not shuffled, the timer may be unexpected by the user if they encounter other shuffled timed trades without timers. Also, there is a special case for Eyedrops not shuffled, but Prescription and Eyeball Frog are. If you find Eyeball Frog, turn it in, and fail the Eyedrops timer, you would revert to Prescription before you find it wherever it was shuffled. The new flags would prevent softlocking on an otherwise unobtainable Eyeball Frog (just d-pad it back to your inventory). You could make it so each timer chain has to be completely unshuffled to activate, or item reverts only go back to the earliest owned item, but trade timers already have enough exceptions to them with ER that I don't want to make this any more complicated than it needs to be from a UX perspective.

> Don't you need to thaw King Zora to trade the Frog?

Yes, thanks for the catch!

> But in general I worry the two logics for rework and no rework are built on top of each other in a kind of awkward way. What would combining them together look like? I hope the 'redundancy?' doesn't cause any odd problems. Not sure what I'm trying to say here.

The parallel logic paths are kept so that we don't need to check the trade quest shuffle setting in the logic strings for reverting items. The non-reverting items/events I kept because I usually screw something up whenever I touch logic, so I try not to anymore.

There was a bug here for Claim Check trade-in where the two paths converge. I added a check for trade shuffle to disable the event path if trade shuffle is on. With trade shuffle on, all adult trade items are in the item pool, either shuffled or in their vanilla locations, so DMT Biggoron will still be reachable logically.

r0bd0g commented:
We don't usually bracket off nots by themselves.

To be clear what I was suggesting was an intermediate setting where somewhere in the world you find prescription (or maybe eyeball frog or drops that could revert to prescription if you don't handle it before saving), and then you would turn script for frog, deliver it on a timer turn it in for drops, and deliver that to Biggoron for the Eyedrops turn in check. (And something similar for turning in Cojiro for mushroom, deliver for odd potion check.) You offered a pretty complicated explanation for why you opted not to do something like this and so I wasn't 100% sure we were on the same page what I was talking about.

mracsys committed:
Full Adult and Child Trade Shuffle

mracsys committed:
Ranch and Saria logic fixes

mracsys committed:
glitched logic typo

mracsys committed:
enable ludicrous pool trade items

ETR-BTF commented:
> So I might have found a bug or something. I have the disable item age checks setting turned on but I imagine this would be an issue for equip swap users too. I also have trade quest shuffle on. I handed the carpenter boss the poacher's saw as child and he gave me the broken goron sword. This first struck me as weird since that's vanilla but then when I switched the trade item to something else I had, the broken goron sword was no longer in my inventory

https://discord.com/channels/274180765816848384/274186223147417600/1037126701282504784

mracsys has already been made aware of this.

fenhl commented on HintList.py:
```diff
@@ -363,7 +363,16 @@ def tokens_required_by_settings(world):
+    'Goron Mask':                                               (["the Goron Mask"],"the Goron Mask",'item'),
+    'Zora Mask':                                                (["the Zora Mask"],"the Zora Mask",'item'),
+    'Gerudo Mask':                                              (["the Gerudo Mask"],"the Gerudo Mask",'item'),
+    'Mask of Truth':                                            (["the Mask of Truth"],"the Mask of Truth",'item'),
```
The new items need cryptic hint text.

ETR-BTF replied:
Gerudo Mask -> Feminine Facade?

ETR-BTF replied:
Zora Mask -> Fishy Facade?

ETR-BTF replied:
Mask of Truth -> Factual or Faithful Facade?

ETR-BTF replied:
Goron Mask -> Familial Facade? Because they call Link brother. That one is quite a stretch tho lol.

fenhl replied:
I like the theme. There's also Chicken, Keaton Mask, Skull Mask, Spooky Mask, and Bunny Hood.

fenhl replied:
The Chicken could just be “a chicken dilemma”, same as the Weird Egg. They're logically equivalent anyway.

fenhl replied:
Spooky Mask could be Frightening Facade.

fenhl replied:
Bunny Hood → Fast Facade

fenhl replied:
Keaton Mask → Famous Facade

fenhl replied:
Skull Mask → Forest Facade? Bit of a stretch.

fenhl replied:
Maybe fraternal for Goron Mask?

ETR-BTF replied:
Fraternal is much better!

ETR-BTF replied:
Fatal Facade for Skull Mask? Because skulls are associated with death? It works better for MM's Blast Mask tho lol.

ETR-BTF replied:
Fleshless? Theatrical?

mracsys committed:
fix logic from trade quest rebase

mracsys committed:
fix adult starting trade items and mask shop

mracsys committed:
carpenter boss trade with equip swap

mracsys committed:
add child mask vague hints

mracsys committed:
shuffle granny's potion shop item

mracsys commented:
This still needs unit tests, but I believe the changes are ready for review as-is. Everything is feature-complete.

mracsys committed:
add trade shuffle development notes

mracsys committed:
granny shuffle logic bug fix

fenhl commented:
For consistency with existing cryptic hints, the new names should be in lowercase, not title case, and should use “a”, not “the”.

mracsys commented:
> Do you actually need to be adult to buy the potion? (You have to be there to trade it, but you don't necessarily have to trade it if the seed has no odd potion.)

You do not need to be adult to buy, that could be a here() to turn in the mushroom. Is there a situation that could matter logically? In gameplay you might not have the wallet as adult and decide to route back as child, but an event wouldn’t guarantee child access in ER as far as I know.

r0bd0g commented:
If the potion shop is only accessible as child, and you don't need to turn the odd potion in to enable the blue potion check. Currently if you start with a trade item beyond odd potion, you can immediately buy. I'm not sure if this PR makes that situation impossible or not, I doubt it, but I dunno.

mracsys commented:
It does handle making blue potion purchasable if you can’t access odd mushroom. I’ll update the logic to handle it.

mracsys committed:
closed kak gate fix and minor cleanup

mracsys committed:
style fixes for CI

cjohnson57 commented:
PR is currently waiting on a fix for right side mask shop items

mracsys committed:
fix mask shop hint preventing purchases

mracsys committed:
use 16bit item IDs for trade functions

mracsys committed:
CI and preset fixes for trade shuffle

mracsys commented:
The right side mask issue is resolved. The misc. hint for those masks had a change to prevent buying them until the trade quest was done, but prevented buying them at all.

My SD card for my Wii isn't working anymore. If anyone has time I'd appreciate tests on that platform.

ETR-BTF commented:
Tested on Wii VC and it seems to work fine! Although I was able to see the bottom-right item on the right hand side even with the misc. hint set to off.

It should be noted that the bottom-right item on the right hand side is given to you automatically upon completing the mask quest. I'm not sure if this is desirable behavior, because the item could mess with equip swap or if it's an Ice Trap it could kill Link on OHKO, making a permadeath/deathless playthrough of such seeds impossible.

cjohnson57 commented:
> making a permadeath/deathless playthrough of such seeds impossible.

Isn't this already the case for ice traps in places like chests? I don't think deathless challenges are something we need to consider for dev

mracsys commented:
> Although I was able to see the bottom-right item on the right hand side even with the misc. hint set to off.

This is vanilla behavior, easy to miss since the curtain hides it.

fenhl commented:
There are already situations where a check can be unavoidable even if you have the spoiler log, for example if HBA 1000 is an ice trap and 1500 is required.

cjohnson57 reviewed (approved):
Amazing work, couldn't find anything to complain about. I triple checked the plando code and it seems good to me.

cjohnson57 commented:
I guess the one very minor issue I have is that this error that displays if you try to start with both pocket egg and pocket cuccoo (and ditto for weird egg/chicken) could be more clear about the cause and how to resolve it but that's definitely not a blocking issue
![image](https://user-images.githubusercontent.com/30701749/218281031-4bef1c9e-d5b2-4c9f-a429-9ed370fbaab7.png)

mracsys committed:
better error messages for chicken/egg paradox
